sisyphus/localengine.py

```diff
@@ -282,23 +307,19 @@ def try_to_recover_task(self, task, task_id):
             rqmt = d['requested_resources']
             logpath = os.path.relpath(task.path(gs.JOB_LOG_ENGINE))
             call_with_id = task.get_worker_call(task_id)
-            call_with_id += ['--redirect_output']
```

Comment:
this arg is actually removed in the worker, so no need to add it here
https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/worker.py#L195

Reply:
There is a whole block of code around that line which will only be executed if '--redirect_output ' is set, only one step is to remove the argument. Starting here:
https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/worker.py#L185

You're just disabling the redirection of stdout and stderr towards the log file, I doubt that you want to do that.

Reply:
I think this is a misunderstanding. Here I remove this in the `try_to_recover_task` function where the `call_with_id` is only created to compare to the original call.

The original call still contains the `--redirect_output` flag. But in the worker, after handling the flag in the code you mentioned, it is removed before the subprocess is spawned. https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/worker.py#L195 Therefore it is not part of the args that the subprocess knows about and must not be added here.

Reply:
Ah, thought this was used to create the worker job. There should be two worker jobs if redirect_output is enabled, one that only redirects the output to the log file (containing the redirect_output flag) and one that does the actual work (doesn't have the redirect_output). 
I thought we are matching against the one with the redirect_output flag, but I was probably wrong about that. The process_id is read from the process logging file (<job_dir>/usage.*) which should be created by the process that doesn't have the redirect_output flag. On the other hand I'm fairly sure it worked at some point 🤔

Reply:
>  There should be two worker jobs if redirect_output is enabled [...] I thought we are matching against the one with the redirect_output flag

Yes and no. In the localengine, when the task is first executed, the PID of the worker with redirect_output flag is stored. cf
https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/localengine.py#L196-L197
https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/localengine.py#L119-L123

But then in `try_to_recover_task` we read the pid from the PLOGGING file
https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/localengine.py#L276-L278
which will be written from the worker that actually runs the task, i.e. without the redirect_output and with the workdir set to the Jobs workdir.
cf. [`LoggingThread.log_usage`](https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/worker.py#L84) as created in [`worker_helper`](https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/worker.py#L233C5-L233C20) in the part that does not have the redirect_output flag

So currently the PID in `running_tasks[name]` changes when the manager is restarted. (With and without my changes)

> On the other hand I'm fairly sure it worked at some point 🤔

Maybe this was before the redirect_output was introduced in https://github.com/rwth-i6/sisyphus/commit/c4f825b043bfd81c75d40f7fd9719e7ffb375f42? after a brief glance in the state after this commit I still see the same code path as indicated above..


Aside from the historical considerations how the current state came to be, do you agree that the current state is not working as intended (Tasks are never recovered) and that this should be fixed?
And do you find my proposed implementation a suitable fix for the problem? 
Or should we rather try to recover the original PID of the process with redirect_output flag?

Reply:
> Maybe it would be good to change the behaviour if a mismatched job is found. Keep the warning that the job changed, but still accept it as the correct job instead of ignoring it.

done now.